Docs update (#73)
solidity-labs-io/forge-proposal-simulator: docs/overview/architecture/proposal-functions.md

> The [Proposal.sol](../../../src/proposals/Proposal.sol) file contains a set of functions that all governance models inherit. There are three governance models: Bravo, Multisig and Timelock already implemented in the example repository. To create a Proposal with FPS, you can inherit one of the governance types, such as [MultisigProposal.sol](../../../src/proposals/MultisigProposal.sol) and override the needed functions. Alternatively, you can create a new governance model in your repository that inherits the `Proposal` contract and then a proposal can be created using a newly formed governance model not already inside the repo. The following is a list of functions all proposals must implement:

Reviewer: ```suggestion
The [Proposal.sol](../../../src/proposals/Proposal.sol) file contains a set of functions that all governance models inherit. There are currently four governance models inheriting from Proposal.sol: Bravo, Multisig, Timelock and OpenZeppelin Governor. When using FPS for any of the abovementioned models for creating proposals, all that needs to be done is to inherit one of the proposal types, such as [GovernorBravoProposal.sol](../../../src/proposals/GovernorBravoProposal.sol) and override the necessary functions to create the proposal, like `build` and `deploy`. FPS is flexible enough so that for any different governance model, governance proposal types can be easily adjusted to fit into the governance architecture. An example has been provided using Arbitrum Governance on FPS example repo (link here) to demonstrate FPS flexibility. The following is a list of functions proposals can implement:
```

-   `function name() public` Override this function to define the proposal name. This function should be overridden at the final proposal level.

-   `function description() public` Override this function to define the proposal description. This function should be overridden at the final proposal level.

> -   `function deploy() public`: Defines new contract deployments. Newly deployed contracts must be added to the `Addresses` contract instance through the setters methods. This function should be overridden only when there are deployments to be made in the proposal.

Reviewer: ```suggestion
-   `function deploy() public`: Defines new contract deployments. Newly deployed contracts must be added to the `Addresses` contract instance through the setters methods. This function should be overridden only when there are deployments to be made in the proposal. The contracts can be broadcasted to the network using the foundry `--broadcast` flag combined with the `--wallet` flag (see https://book.getfoundry.sh/reference/cast/cast-wallet for reference). A `DEPLOYER_EOA` address must exist in Addresses.json file. 
```

> -   `function afterDeployMock() public`: Specifies post-deployment actions. Such actions can include wiring contracts together, transferring ownership rights, or invoking setter functions as the deployer.

Reviewer: ```suggestion
-   `function afterDeployMock() public`: Post-deployment mock actions. Such actions can include pranking, etching, etc.
```

> -   `function build() public`: Creates the proposal actions and saves them to storage in the proposal contract.

Reviewer: ```suggestion
-   `function build() public`: Creates the proposal actions and saves them to storage for later usage on `simulate` and `getCalldata` functions. It's mandatory to use the `buildModifier` with the contract address that will perform the actions as a parameter, e.g., the timelock address.
```

> -   `function build() public`: Creates the proposal actions and saves them to storage in the proposal contract.

Reviewer: missing `_validateActions` and `_validateAction` functions

Author: Added

> -   `function simulate() public`: Executes the actions that were previously saved during the `build` step. This function's execution is dependent on the successful execution of the `build` function. Without calling `build` first, the `_simulateActions()` function becomes ineffectual as there would be no predefined actions to execute.

Reviewer: ```suggestion
-   `function simulate() public`: Executes the previously saved actions during the `build` step. This function's execution is dependent on the successful execution of the `build` function. Without calling `build` first, `simulate` function becomes ineffectual as there would be no predefined actions to execute. 
```

> -   `function run() public`: Simulates proposal execution against a forked mainnet or locally and can be used in integration tests or scripts. This function can be overridden at the final proposal level where it can be used to set the environment, for eg: addresses object, primary fork id. Make sure to call the `run()` from the Proposal contract as it performs key actions.

Reviewer: ```suggestion
-   `function run() public`: Simulates the execution of a proposal against a forked mainnet or locally, and it can be used in integration tests or scripts. It must be overridden at the final proposal level to set up the environment, including creating the addresses instance, setting the governance contract, and creating the primary fork. When overriding, make sure to call `super.run()`.
```

> -   `function checkOnChainCalldata() public`: Check if there are any on-chain proposal that matches the proposal calldata. This function should be overriden at the governance contract level. [Timelock Proposal](../../../src/proposals/TimelockProposal.sol), [Governor Bravo Proposal](../../../src/proposals/GovernorBravoProposal.sol)

Reviewer: ```suggestion
-   `function checkOnChainCalldata() public`: Check if any on-chain proposal matches the proposal calldata. There is no need to override this function at the proposal contract level as it is already overridden in the proposal type contract. 
```

> -   `function print() public`: Print proposal description, actions and calldata

Reviewer: ```suggestion
-   `function print() public`: Print proposal description, actions and calldata. No need to override. 
```

-   `function getProposalActions() public`: Retrieves the sequence of actions for a proposal. This function should not be overridden in most of the cases.

> -   `function getCalldata() public`: Retrieves any generated governance proposal calldata. This function should be overriden at the final proposal level.

Reviewer: ```suggestion
-   `function getCalldata() public`: Retrieves any generated governance proposal calldata. TainCalldata() public`: Check if any on-chain proposal matches the proposal calldata. There is no need to override this function at the proposal contract level as it is already overridden in the proposal type contract.
```

-   `function validate() public`: Validates system state post-execution. This allows checking that contract's variables and newly deployed contracts are set up correctly.

The actions in FPS are designed to be loosely coupled for flexible

> implementation, with the exception of the build and run functions, which require

Reviewer: ```suggestion
implementation, with the exception of the build and simulate functions, which require
```

sequential execution. This design choice offers developers significant
flexibility and power in tailoring the system to their specific needs. For
example, a developer may choose to only execute the deploy and validate
functions, bypassing the others. This could be suitable in situations where only
initial deployment and final validation are necessary, without the need for
intermediate steps. Alternatively, a developer might opt to simulate a proposal

> by executing only the build and run functions, omitting the deploy step if there

Reviewer: ```suggestion
by executing only the build and simulate functions, omitting the deploy step if there
```

Author: Done

> The [Proposal.sol](../../../src/proposals/Proposal.sol) file contains a set of functions that all governance models inherit. Currently, there are four governance models inheriting from Proposal.sol: Bravo, Multisig, Timelock, and OpenZeppelin Governor. When using FPS for any of the aforementioned models to create proposals, all that needs to be done is to inherit one of the proposal types, such as [GovernorBravoProposal.sol](../../../src/proposals/GovernorBravoProposal.sol), and override the necessary functions to create the proposal, like `build` and `deploy`. FPS is flexible enough so that for any different governance model, governance proposal types can be easily adjusted to fit into the governance architecture. An example has been provided using Arbitrum Governance on the FPS example repo (link here) to demonstrate FPS flexibility. The following is a list of functions proposals can implement:

Reviewer: replace (link here) for actual link

-   `function name() public`: This function is empty in the `Proposal` contract.

    ```solidity
    function name() external view virtual returns (string memory);

>     ```

Reviewer: ```suggestion
-   `function name() external view virtual returns (string memory);`: This function is empty in the `Proposal` contract.
```

remove the duplicated information for this function and all below.

Author: The two name are different. One is from `Proposal` contract and another is the overridden function in the proposal specific contract. @anajuliabit

Reviewer: When writing a sentence, consider what message you're trying to convey to the user. If you think showing the user the interface is important, then use only the interface method. They are nearly identical, so there's no need to provide both. Let's avoid including unnecessary information. If you've already provided the information, there's no need to duplicate it.

Author: I am trying to follow a pattern where I explain the user how the functions are defined at the proposal contract level and then overridden later in the proposal specific contract and I am following this pattern for all the methods. I think showing user both the proposal contract implementation and the overridden implementation is good for a new user although they are nearly identical for some of the functions.

Reviewer: ![Screenshot 2024-06-02 at 19 38 29](https://github.com/solidity-labs-io/forge-proposal-simulator/assets/50335980/cb9dd3ef-ae86-4da1-af4c-1a5aa4870a2c)

Take for example the function `name`. You wrote three solidity-snipped codes. That's just make things complicated  instead of helping with clarification 

@ElliotFriedman what do you think?

Reviewer 2: Looks like it is simplified now

    function name() public pure override returns (string memory) {
        return "BRAVO_MOCK";
    }
    ```

-   `description()`: This function is empty in the Proposal contract. Override this function in the proposal-specific contract to define the proposal description. For example:

    ```solidity
    function description() public pure override returns (string memory) {
        return "Bravo proposal mock";
    }
    ```

> -   `deploy()`: This function is empty in the Proposal contract. Override this function when there are deployments to be made in the proposal. Here is an example from a [governor bravo proposal](https://github.com/solidity-labs-io/forge-proposal-simulator/blob/6815ce32b86b60cb4bec7d3cb34ad9f15cbbe8dd/docs/guides/governor-bravo-proposal.md) demonstrating how to deploy two contracts, Vault and Token, if they are not already deployed.

Reviewer 2: ```suggestion
-   `deploy()`: This function is empty in the Proposal contract. Override this function when there are deployments to be made in the proposal. Here is an example from a [Governor Bravo Proposal](https://github.com/solidity-labs-io/forge-proposal-simulator/blob/6815ce32b86b60cb4bec7d3cb34ad9f15cbbe8dd/docs/guides/governor-bravo-proposal.md) demonstrating how to deploy two contracts, Vault and Token, if they are not already deployed.
```

    Override this function in the proposal-specific contract to define the proposal description. For example:

    ```solidity
    function description() public pure override returns (string memory) {
        return "Bravo proposal mock";
    }
    ```

-   `function deploy() public`: This function is empty in the `Proposal` contract.

    ```solidity
    function deploy() public virtual {}
    ```

>     Override this function when there are deployments to be made in the proposal. Here is an example from a [governor bravo proposal](../../guides/governor-bravo-proposal.md). Here two contracts are deployed, Vault and Token if they are not already deployed.

Reviewer: ```suggestion
    Override this function when there are deployments to be made in the proposal. Here is an example from a [governor bravo proposal](../../guides/governor-bravo-proposal.md) demonstrating how to deploy two contracts, Vault and Token, if they are not already deployed.
```

            addresses.addAddress("BRAVO_VAULT_TOKEN", address(token), true);
            token.transferOwnership(owner);

            uint256 balance = token.balanceOf(address(this)) > 0
                ? token.balanceOf(address(this))
                : token.balanceOf(addresses.getAddress("DEPLOYER_EOA"));

            token.transfer(address(owner), balance);
        }
    }
    ```

>     Deployments are done by `DEPLOYER_EOA` when running through a proposal `forge script,` and therefore, an address with this exact name must exist in the Addresses.json file. Foundry can be leveraged to actually broadcast the deployments when using the `broadcast` flag combined with `--account` flag. Please refer to the [foundry docs](https://book.getfoundry.sh/tutorials/best-practices?highlight=broadcast#scripts) for further assistance. Alternatively, when proposals are executed through `forge script,` the deployer address is the proposal contract itself.

Reviewer 2: ```suggestion
    Deployments are done by `DEPLOYER_EOA` when running through a proposal `forge script,` and therefore, an address with this exact name must exist in the Addresses.json file. Foundry can be leveraged to actually broadcast the deployments when using the `broadcast` flag combined with `--account` flag. Please refer to the [foundry docs](https://book.getfoundry.sh/tutorials/best-practices?highlight=broadcast#scripts) for more information. Alternatively, when proposals are executed through `forge script,` the deployer address is the proposal contract itself.
```

        if (!addresses.isAddressSet("BRAVO_VAULT_TOKEN")) {
            // Here the token is deployed
            Token token = new Token();
            addresses.addAddress("BRAVO_VAULT_TOKEN", address(token), true);
            token.transferOwnership(owner);

            uint256 balance = token.balanceOf(address(this)) > 0
                ? token.balanceOf(address(this))
                : token.balanceOf(addresses.getAddress("DEPLOYER_EOA"));

            token.transfer(address(owner), balance);
        }
    }
    ```

>     Deployments are done by `DEPLOYER_EOA` address defined in `Addresses.json`. We get this address while setting the proposal environment in the `run` function; please refer to the `run` function documentation below to get more details. In tests, `deploy()` is used to simulate the proposal contract deployments, and on the other hand, when the proposal is run as a script with the broadcast flag and deployer EOA private key, it is used to deploy all the contracts on the chain and adds those addresses to the `addresses` object.

Reviewer: ```suggestion
    Deployments are done by `DEPLOYER_EOA` when running through a proposal `forge script,` and therefore, an address with this exact name must exist in the Addresses.json file. Foundry can be leveraged to actually broadcast the deployments when using the `broadcast` flag combined with `--account` flag. Please refer to the [foundry docs[(https://book.getfoundry.sh/tutorials/best-practices?highlight=broadcast#scripts) for further assistance. On the other hand, when proposals are executed through a `forge script,` the deployer is the proposal contract itself.
```

-   `function afterDeployMock() public`: Post-deployment mock actions. Such actions can include pranking, etching, etc.

<a id="#build-function"></a>

> -   `function build() public`: Besides deployments, a proposal will have some actions to be made, for example, transferring ownership rights of a deployed contract. This function helps create and store these actions, with each action storing a target contract and calldata for executing that action. In the Proposal contract, it is an empty function. This function should always be overridden in the proposal-specific contract, as every proposal must have at least a single action to be made.

Reviewer: ```suggestion
-   `function build() public`: this function is where most of the FPS magic happens. It utilizes foundry cheat codes to automatically transform plain solidity code into calldata. This calldata can then be proposed on the Governance, signed by the multisig signers, or scheduled on the Timelock. For instance, an action might involve pointing a proxy to a new implementation address after deploying the implementation in the 'deploy' function.
```

    ```solidity
    function build() public virtual {}

>     ```

Reviewer: ```suggestion
```

>     In the `startBuild()` function, set the prank to the caller address and take a snapshot of the initial state using Foundry's `vm.snapshot()` cheat code. Then, initiate Foundry's `vm.startStateDiffRecording()` to start recording all function calls made after this step. The `build()` function is then executed, and all the build steps are recorded. Finally, `endBuild()` stops the state diff recording, retrieves all the call information, stops the prank, reverts the state to the initial snapshot, and filters out the calls made by the caller, ignoring any static calls. This process ensures that only the mutative calls made by the caller are filtered and stored in the actions array.

Reviewer 2: ```suggestion
    The `startBuild()` function pranks as the caller address, and takes a snapshot of the initial state using Foundry's `vm.snapshot()` cheat code. Then, it initiates a call to Foundry's `vm.startStateDiffRecording()` cheatcode to record all function calls made after this step. The `build()` function is then executed, and all the build steps are recorded. Finally, the `endBuild()` function is called which stops the state diff recording, retrieves all the call information, stops the prank and reverts the state to the initial snapshot. It then filters out the calls to only ones made by the specified caller address, ignoring any static calls and calls made to Foundry's Vm contract. This process ensures that only mutative calls made by the caller are filtered and stored in the actions array.
```

>     This function helps the user store the actions without having to write any complicated calldata generation code by cleverly using a few Foundry features. This can be best understood by looking at the `buildModifier` in the Proposal contract. This modifier runs the `_startBuild()` function before `build()` and the `endBuild()` function after `build()`. `toPrank` is also passed as a parameter to this modifier, which is the address that will be used as the caller for the actions in the proposal, e.g., multisig address, timelock address, etc. In the `startBuild()` function, first of all, we prank to the caller address and then take a snapshot of the start state by using Foundry's `vm.snapshot()` feature. After this, we call Foundry's `vm.startStateDiffRecording()`, which will now start recording all the function calls made after this step. Next, the `build()` function will run, and all the steps in the build will be recorded. Finally, `endBuild()` will run; first, it will stop the state diff recording and get all the call info, then stop the prank, then revert the state to the start snapshot, and finally, it will filter the calls made by the caller and also ignore the static calls. This way, only the mutative calls made by the caller will be filtered out, and these calls are finally stored in the actions array. Inside end build, duplicate actions are also checked using `_validateAction`, and custom checks can also be created using `_validateActions`, like the target should always be the timelock contract, value should always be 0, or any custom action check.

Reviewer: ```suggestion
  The `buildModifier` must be implemented when overriding so that FPS can store the actions to be used later on in the calldata generation. This modifier runs the `_startBuild()` function before `build()` and the `endBuild()` function after.  This modifier also takes `toPrank` as a parameter, which represents the address used as the caller for the actions in the proposal, such as the multisig address or timelock address.

In the `startBuild()` function, we set the prank to the caller address and take a snapshot of the initial state using Foundry's `vm.snapshot()` cheat code. Then, we initiate Foundry's `vm.startStateDiffRecording()` to start recording all function calls made after this step. The `build()` function is then executed, and all the build steps are recorded. Finally, `endBuild()` stops the state diff recording, retrieves all the call information, stops the prank, reverts the state to the initial snapshot, and filters out the calls made by the caller, ignoring any static calls. This process ensures that only the mutative calls made by the caller are filtered and stored in the actions array.```

validateActions and validateAction should be explained each one in its section

>     ///  3). starting a $recording of all calls created during the proposal

Reviewer 2: ```suggestion
    ///  3). starting a recording of all calls created during the proposal
```

                accountAccesses[i].accessor == caller /// caller is correct, not a subcall
            ) {
                _validateAction(
                    accountAccesses[i].account,
                    accountAccesses[i].value,
                    accountAccesses[i].data
                );

                actions.push(
                    Action({
                        value: accountAccesses[i].value,
                        target: accountAccesses[i].account,
                        arguments: accountAccesses[i].data,
                        description: string(
                            abi.encodePacked(
                                "calling ",
                                vm.toString(accountAccesses[i].account),
                                " with ",
                                vm.toString(accountAccesses[i].value),
                                " eth and ",
                                vm.toString(accountAccesses[i].data),
                                " data."
                            )
                        )
                    })
                );
            }

            _validateActions();
        }
    }
    ```

>     For a clearer understanding, take a look at the Bravo proposal example snippet. Here, `build()` has the `buildModifier`, which takes `PROTOCOL_TIMELOCK_BRAVO` as the caller. Multiple calls are made in the `build()` function, but only the mutative calls made by the caller will be stored in the actions array. In this example, there are three mutative actions. First, the deployed token is whitelisted on the deployed vault. Second, the vault is approved to transfer some `balance` on behalf of the Bravo timelock contract. Third, the `balance` amount of tokens is deposited in the vault from the Bravo timelock contract.

Reviewer 2: ```suggestion
    For a clearer understanding, take a look at the Bravo proposal example snippet. Here, `build()` has the `buildModifier`, which takes the address of `PROTOCOL_TIMELOCK_BRAVO` as the caller. Multiple calls are made in the `build()` function, but only the mutative calls made by the caller will be stored in the actions array. In this example, there are three mutative actions. First, the deployed token is whitelisted on the deployed vault. Second, the vault is approved to transfer some `balance` on behalf of the Bravo timelock contract. Third, the `balance` amount of tokens is deposited in the vault from the Bravo timelock contract.
```

    ```solidity
    function build()
        public
        override
        buildModifier(addresses.getAddress("PROTOCOL_TIMELOCK_BRAVO"))
    {
        /// STATICCALL -- non-mutative and hence not recorded for the run stage

        // Get vault address.
        address bravoVault = addresses.getAddress("BRAVO_VAULT");

        // Get token address.
        address token = addresses.getAddress("BRAVO_VAULT_TOKEN");

        // Get Bravo timelock's token balance.
        uint256 balance = Token(token).balanceOf(
            addresses.getAddress("PROTOCOL_TIMELOCK_BRAVO")
        );

        /// CALLS -- mutative and recorded

        // Whitelists the deployed token on the deployed vault.
        Vault(bravoVault).whitelistToken(token, true);

        // Approve the token for the vault.
        Token(token).approve(bravoVault, balance);

        // Deposit all tokens into the vault.
        Vault(bravoVault).deposit(token, balance);
    }
    ```

> -   `_validateAction()`: It ensures there are no duplicate actions. This method can be further overridden to include custom checks for a action.

Reviewer 2: ```suggestion
-   `_validateAction()`: Hook that reviews a single proposal action. In this example, it ensures the action being added to the proposal is not a duplicate. This method can be further overridden to include additional checks for all proposal actions.
```

    ```solidity
    function _validateAction(
        address target,
        uint256 value,
        bytes memory data
    ) internal virtual {
        uint256 actionsLength = actions.length;
        for (uint256 i = 0; i < actionsLength; i++) {
            // Check if the target, arguments and value matches with other exciting actions.
            bool isDuplicateTarget = actions[i].target == target;
            bool isDuplicateArguments = keccak256(actions[i].arguments) ==
                keccak256(data);
            bool isDuplicateValue = actions[i].value == value;

            require(
                !(isDuplicateTarget &&
                    isDuplicateArguments &&
                    isDuplicateValue),
                "Duplicated action found"
            );
        }
    }
    ```

> -   `_validateActions()`: This method can be overridden to add custom checks on all actions of proposals. For example, there must be a single action, the first action should be an approval call, etc.

Reviewer 2: ```suggestion
-   `_validateActions()`: This method checks all proposal actions based on user defined checks. It can be overridden to add custom checks on all actions of proposals. As an example, checks performed could be that if a certain function is called, the next three actions should be a predefined set of actions, and if they are not, then revert.
```

> -   `getProposalActions()`: Retrieves the sequence of actions for a proposal. This function should not be overridden in most cases.

Reviewer 2: What is its interface?

Author: what do u mean by this? can u please add more details. @ElliotFriedman

Reviewer 2: what it returns

Reviewer 2: it's important to define the full interface.

you do the same for the name, symbol, deploy, afterDeployMock and run.

it should be the same here

Reviewer 2: you need to include return types for all of these functions by showing the user the interface:

- _validateActions
- getProposalActions
- getCalldata

-   `getCalldata()`: Retrieves any generated governance proposal calldata. This function should not be overridden at the proposal contract level as it is already overridden in the proposal type contract.

<a id="#run-function"></a>

-   `run()`: This function serves as the entry point for proposal execution. It selects the `primaryForkId` which will be used to run the proposal simulation. It executes `deploy()`, `afterDeployMock()`, `build()`, `simulate()`, `validate()`, and `print()` in that order if the flag for a function is set to true. `deploy()` is encapsulated in start and stop broadcast. This is done so that contracts can be deployed on-chain.

    ```solidity
    function run() public virtual {
        vm.selectFork(primaryForkId);

        if (DO_DEPLOY) {
            address deployer = addresses.getAddress("DEPLOYER_EOA");

            vm.startBroadcast(deployer);
            deploy();
            addresses.printJSONChanges();
            vm.stopBroadcast();
        }

        if (DO_AFTER_DEPLOY_MOCK) afterDeployMock();
        if (DO_BUILD) build();
        if (DO_SIMULATE) simulate();
        if (DO_VALIDATE) validate();
        if (DO_PRINT) print();
    }
    ```

>     This function is overridden at the proposal-specific contract. For example, let's take a look at the Governor Bravo `run()` method. Here, this function sets the environment for proposal execution and then finally simulates the proposal by calling the `run()` of the parent `Proposal` contract. In this example, first `primaryForkId` is set to `sepolia`. Next, the `addresses` object is set by reading the `Addresses.json` file. The Bravo governor specific contract requires setting the Bravo governor contract address as it is required in `simulate()` and `checkOnChainCalldata()` functions. Finally, `run()` sets the Bravo governor address and calls `super.run()`.

Reviewer 2: This is a pretty wordy paragraph. Would be great if it could be made more succinct

>     This function is overridden at the proposal-specific contract. Let's take a look at the Governor Bravo `run()` method. Here, this function sets the environment for proposal execution and then finally simulates the proposal by calling the `run()` of the parent `Proposal` contract using `super.run()`. In this example, first `primaryForkId` is set to `sepolia`. Next, the `addresses` object is set by reading the `Addresses.json` file. Finally `governor` is set using `setGovernor()` method which is required in `simulate()` and `checkOnChainCalldata()` functions.

Reviewer 2: we should use less casual and more of a formal tone here "Let's take a look at the Governor Bravo ".

Something more along the lines of:

"Looking at the run function in governor bravo explains how this function works."

    ```solidity
    function run() public override {
        // Create and select sepolia fork for proposal execution.
        primaryForkId = vm.createFork("sepolia");
        vm.selectFork(primaryForkId);

        // Set addresses object reading addresses from json file.
        setAddresses(
            new Addresses(
                vm.envOr("ADDRESSES_PATH", string("addresses/Addresses.json"))
            )
        );

        // Set governor Bravo. This address is used for proposal simulation and check on
        // chain proposal state.
        setGovernor(addresses.getAddress("PROTOCOL_GOVERNOR"));

        // Call the run function of parent contract 'Proposal.sol'.
        super.run();
    }
    ```

> -   `simulate()`: Executes the previously saved actions during the `build` step. This function's execution depends on the successful execution of the `build` function.

Reviewer 2: what is the difference between simulate and run?

Author: run() is the entry point that runs all other functions including simulate(). simulate() simulates proposal actions that were added in the build() step.

        // Warp to allow proposal execution on timelock
        ITimelockBravo timelock = ITimelockBravo(governor.timelock());
        vm.warp(block.timestamp + timelock.delay());

        // Execute the proposal
        governor.execute(proposalId);
        require(
            governor.state(proposalId) == IGovernorBravo.ProposalState.Executed
        );
    }
    ```

>     Without calling `build` first, the `simulate` function becomes ineffectual as there would be no predefined actions to execute.

Reviewer 2: ```suggestion
    Without calling `build` first, the `simulate` function would be a no-op as there would be no predefined actions to execute.
```

>     This function can be overridden at the governance-specific contract or the proposal-specific contract, depending on the type of proposal. For example, the Governor Bravo type overrides this function at the governance-specific contract, while the Timelock type overrides it at the proposal-specific contract. Take a look at the Governor Bravo type example. In this example, the proposer address proposes the proposal to the governance contract. It first transfers and delegates governance tokens that meet the minimum proposal threshold and quorum votes to itself. Then it registers the proposal, rolls the proposal to the Active state so that voting can begin, votes yes to the proposal, rolls to the block where the voting period has ended, and now the proposal is in the Succeeded state. Then it queues the proposal in the governance Timelock contract, warps to the end of the timelock delay period, and finally executes the proposal, thus simulating the complete proposal on the local fork.

Reviewer 2: ```suggestion
    This function can be overridden at the governance-specific contract or the proposal-specific contract, depending on the type of proposal. For example, the Governor Bravo type overrides this function at the governance-specific contract, while the Timelock type overrides it at the proposal-specific contract. In the Governor Bravo example, the proposer address proposes the proposal to the governance contract. It first transfers and delegates governance tokens that meet the minimum proposal threshold and quorum votes to itself. Then it registers the proposal, rolls the proposal to the Active state so that voting can begin, votes yes to the proposal, rolls to the block where the voting period has ended, putting the proposal in the Succeeded state. Then it queues the proposal in the governance Timelock contract, warps to the end of the timelock delay period, and finally executes the proposal, thus simulating the complete proposal on the local fork.
```

        require(
            governor.state(proposalId) == IGovernorBravo.ProposalState.Pending
        );

        // Roll to Active state (voting period)
        vm.roll(block.number + governor.votingDelay() + 1);
        require(
            governor.state(proposalId) == IGovernorBravo.ProposalState.Active
        );

        // Vote YES
        vm.prank(proposerAddress);
        governor.castVote(proposalId, 1);

        // Roll to allow proposal state transitions
        vm.roll(block.number + governor.votingPeriod());
        require(
            governor.state(proposalId) == IGovernorBravo.ProposalState.Succeeded
        );

        // Queue the proposal
        governor.queue(proposalId);
        require(
            governor.state(proposalId) == IGovernorBravo.ProposalState.Queued
        );

        // Warp to allow proposal execution on timelock
        ITimelockBravo timelock = ITimelockBravo(governor.timelock());
        vm.warp(block.timestamp + timelock.delay());

        // Execute the proposal
        governor.execute(proposalId);
        require(
            governor.state(proposalId) == IGovernorBravo.ProposalState.Executed
        );
    }
    ```

>     Without calling `build` first, the `simulate` function would be a no-op as there would be no predefined actions to execute.

Reviewer 2: ```suggestion
    Without calling `build` first, the `simulate` function would be a no-op as there would be no defined actions to execute.
```

> -   `validate()`: Validates the system state post-proposal simulation. This allows checking that the contract's variables and newly deployed contracts are set up correctly and that the actions in the build were simulated correctly.

Reviewer 2: ```suggestion
-   `validate()`: Validates the system state post-proposal simulation. This allows checking that the contract's variables and newly deployed contracts are set up correctly and that the actions in the build function were applied correctly to the system.
```

>     This function is overridden at the proposal-specific contract. For example, take a look at the Governor Bravo `validate()` method. It checks the deployment by ensuring the total supply is 10 million and Bravo Timelock is the owner of the deployed token and vault, as the ownership of these contracts was transferred from the deployer EOA to the Bravo Timelock in the `deploy()` method for this proposal. It checks the build actions simulation by ensuring that the token was successfully whitelisted on the vault and the vault's token balance is equal to the Bravo Timelock's deposit in the vault.

Reviewer 2: ```suggestion
    This function is overridden at the proposal-specific contract. For example, take a look at the Governor Bravo `validate()` method. It checks the deployment by ensuring the total supply is 10 million and Bravo Timelock is the owner of the deployed token and vault, as the ownership of these contracts was transferred from the deployer EOA to the Bravo Timelock in the `deploy()` method for this proposal. It checks the proposal actions simulation by ensuring that the token was successfully whitelisted on the vault and the vault's token balance is equal to the Bravo Timelock's deposit in the vault.
```

    ```solidity
    function run() public override {
        // Create and select sepolia fork for proposal execution.
        primaryForkId = vm.createFork("sepolia");
        vm.selectFork(primaryForkId);

        // Set addresses object reading addresses from json file.
        setAddresses(
            new Addresses(
                vm.envOr("ADDRESSES_PATH", string("addresses/Addresses.json"))
            )
        );

        // Make 'addresses' state persist across selected fork.
        vm.makePersistent(address(addresses));

        // Set governor Bravo. This address is used for proposal simulation and check on
        // chain proposal state.
        setGovernor(addresses.getAddress("PROTOCOL_GOVERNOR"));

        // Call the run function of parent contract 'Proposal.sol'.
        super.run();
    }
    ```

> -   `function checkOnChainCalldata() public`: Check if there are any on-chain proposals that match the proposal calldata. There is no need to override this function at the proposal contract level as it is already overridden in the proposal type contract. [Timelock Proposal](../../../src/proposals/TimelockProposal.sol), [Governor Bravo Proposal](../../../src/proposals/GovernorBravoProposal.sol)

Reviewer: ```suggestion
-   `function checkOnChainCalldata() public`: Check if any on-chain proposals match the proposal calldata. There is no need to override this function at the proposal contract level as it is already overridden in the proposal type contract.
```

Author: This was asked by Elliot to be added in one of the previous comments. And I think its good to have references here if someone wants to have a closer look.

Reviewer: then you should add a text describing why those link are useful instead of just dropping the link at the end of the sentence

Author: Done

> -   `checkOnChainCalldata()`: Check if there are any on-chain proposals that match the proposal calldata. There is no need to override this function at the proposal specific contract as it is already overridden in the governance specific contract. Check [Timelock Proposal](../../../src/proposals/TimelockProposal.sol), [Governor Bravo Proposal](../../../src/proposals/GovernorBravoProposal.sol) and [Governor OZ Proposal](../../../src/proposals/GovernorOZProposal.sol) to get implementation details for each proposal tyoe.

Reviewer 2: We should add information about what happens for this method for gnosis safe multisigs here as I believe it is not supported and just reverts (but you'd need to double check this)

> -   `checkOnChainCalldata()`: Check if there are any on-chain proposals that match the proposal calldata. There is no need to override this function at the proposal specific contract as it is already overridden in the governance specific contract. Check [Timelock Proposal](../../../src/proposals/TimelockProposal.sol), [Governor Bravo Proposal](../../../src/proposals/GovernorBravoProposal.sol) and [Governor OZ Proposal](../../../src/proposals/GovernorOZProposal.sol) to get implementation details for each proposal type. This function is not implemented for a multisig proposal as it is not possible to check the state of a gnosis safe transaction on chain. This method reverts with "Not implemented" message for multisig proposals.

Reviewer 2: OZ Governor

-   `function print() public`: Print proposal description, actions, and calldata. No need to override.

> The actions in FPS are designed to be loosely coupled for flexible implementation, with the exception of the build and run functions, which require sequential execution. This design choice offers developers significant flexibility and power in tailoring the system to their specific needs. For example, a developer may choose to only execute the deploy and validate functions, bypassing the others. This could be suitable in situations where only initial deployment and final validation are necessary, without the need for intermediate steps. Alternatively, a developer might opt to simulate a proposal by executing only the build and run functions, omitting the deploy step if there is no need to deploy new contracts. FPS empowers developers with the ability to pick and choose speeds integration tests, deployment scripts, and governance proposal creation as it becomes easy to access whichever part of a governance proposal that is needed.

Reviewer 2: ```suggestion
The actions in FPS are designed to be loosely coupled for flexible implementation, with the exception of the build and run functions, which require sequential execution. This design choice offers developers significant flexibility and power in tailoring the system to their specific needs. For example, a developer may choose to only execute the deploy and validate functions, bypassing the others. This could be suitable in situations where only initial deployment and final validation are necessary, without the need for intermediate steps. Alternatively, a developer might opt to simulate a proposal by executing only the build and run functions, omitting the deploy step if there is no need to deploy new contracts. FPS empowers developers with the ability to pick and choose functions from a proposal for integration tests, deployment scripts, and governance proposal creation as it becomes easy to access whichever part of a governance proposal that is needed exactly how it will be run in production.
```

Author: done

> The actions in FPS are designed to be loosely coupled for flexible implementation, with the exception of the build and simulate functions, which require sequential execution. This design choice offers developers significant flexibility and power in tailoring the system to their specific needs. For example, a developer may choose to only execute the deploy and validate functions, bypassing the others. This could be suitable in situations where only initial deployment and final validation are necessary, without the need for intermediate steps. Alternatively, a developer might opt to simulate a proposal by executing only the build and simulate functions, omitting the deploy step if there is no need to deploy new contracts. FPS empowers developers with the ability to pick and choose functions from a proposal for integration tests, deployment scripts, and governance proposal creation as it becomes easy to access whichever part of a governance proposal that is needed exactly how it will be run in production.

Reviewer 2: ```suggestion
The actions in FPS are designed to be loosely coupled for flexible implementation, with the exception of the build and simulate functions, which require sequential execution. This design choice offers developers significant flexibility and power in tailoring the system to their specific needs. For example, a developer may choose to only execute the deploy and validate functions, bypassing the others. This could be suitable in situations where only initial deployment and final validation are necessary, without the need for intermediate steps. Alternatively, a developer might opt to simulate a proposal by executing only the build and simulate functions, omitting the deploy step if there is no need to deploy new contracts. FPS empowers developers with the ability to pick and choose functions from a proposal for integration tests, deployment scripts, and governance proposal creation as it becomes easy to access whichever part of a governance proposal that is needed and test it locally in as close to production conditions as possible.
```

> The actions in FPS are designed to be loosely coupled for flexible implementation, with the exception of the build and simulate functions, which require sequential execution. This design choice offers developers significant flexibility and power in tailoring the system to their specific needs. For example, a developer may choose to only execute the deploy and validate functions, bypassing the others. This could be suitable in situations where only initial deployment and final validation are necessary, without the need for intermediate steps. Alternatively, a developer might opt to simulate a proposal by executing only the build and simulate functions, omitting the deploy step if there is no need to deploy new contracts. FPS empowers developers with the ability to pick and choose functions from a proposal for integration tests, deployment scripts, and governance proposal creation as it becomes easy to access whichever part of a governance proposal that is needed and test it locally in as close to production conditions as possible.

Reviewer 2: what about build and run?

Author: `build` and `simulate` have to be in a sequence as `simulate` function simulates the actions added in the build step. @ElliotFriedman

> The `Proposal.sol` file contains a set of functions that all governance models inherit. There are three governance models bravo, multisig, timelock already implemented in the repo. To create a Proposal compatible with FPS, you can inherit one of the governance models, such as `MultisigProposal.sol,` and override the needed functions, or you can create a new governance model in your repository that inherits the `Proposal` contract and then a proposal can be created using this new governance model. Following is the list of functions:

Reviewer 2: ```suggestion
The `Proposal.sol` file contains a set of functions that all governance models inherit. There are three governance models: Bravo, Multisig and Timelock already implemented in the example repository. To create a Proposal with FPS, you can inherit one of the governance types, such as `MultisigProposal.sol,` and override the needed functions. Alternatively, you can create a new governance model in your repository that inherits the `Proposal` contract and then a proposal can be created using a newly formed governance model not already inside the repo. The following is a list of functions all proposals must implement:
```

provide links to all three examples

provide link to `MultisigProposal.sol` when you mention

provide link to Proposal contract

> The `Proposal.sol` file contains a set of functions that all governance models inherit. There are three governance models bravo, multisig, timelock already implemented in the repo. To create a Proposal compatible with FPS, you can inherit one of the governance models, such as `MultisigProposal.sol,` and override the needed functions, or you can create a new governance model in your repository that inherits the `Proposal` contract and then a proposal can be created using this new governance model. Following is the list of functions:

Reviewer 2: the ordering and grouping of points in this file is kind of confusing.

The first category of functions should be view only functions

Second should be action steps in the order of execution to make it easy to follow along.

Print and validate going at the end make sense

Author: done

> -   `function description() public` Override this function to define the proposal description. This function should be overriden at the final proposal level.

Reviewer 2: ```suggestion
-   `function description() public` Override this function to define the proposal description. This function should be overridden at the final proposal level.
```

what does "final proposal level" mean? this should be clarified.

Author: It means the Proposal Specific Contract. updating this.

> -   `function getCalldata() public`: Retrieves any generated governance proposal calldata. This function should be overriden at the final proposal level.

Reviewer 2: "final proposal level." needs to be more clearly defined

Author: done

> -   `function checkOnChainCalldata() public`: Check if there are any on-chain proposal that matches the proposal calldata. This function should be overriden at the governance contract level.

Reviewer 2: isn't this already implemented for governor bravo and timelock? let's link out to that here

Author: done

> -   `function run() public`: Simulates proposal execution against a forked mainnet or locally and can be used in integration tests or scripts. This function can be overridden at the final proposal level where it can be used to set the environment, for eg: addresses object, primary fork id. Make sure to call the `run()` from the Proposal contract as it performs key actions.

Reviewer 2: again with the "final proposal level", need to clarify that

Either the description of what is happening here should be expanded on in great detail, or just provide links to examples if you aren't going to bother explaining yourself.

talking about the arbitrum example of the run function would explain a lot of concepts, as you can talk about switching between networks. but you also cannot switch networks in the deploy function as that is not allowed while doing a broadcast.

you need to get really specific here and flesh out all the things that are happening at each level.

e.g. add new contract addresses to the addresses object after deployment (doesn't really fit in run function)

-   `function getProposalActions() public`: Retrieves the sequence of actions for a proposal. This function should not be overridden in most of the cases.

The following functions allow for a significant level of customization and can be overridden in your proposal contract as needed.

-   `function deploy() public`: Defines new contract deployments. Newly deployed contracts must be added to the `Addresses` contract instance through the setters methods. This function should be overridden only when there are deployments to be made in the proposal.

-   `function afterDeployMock() public`: Specifies post-deployment actions. Such actions can include wiring contracts together, transferring ownership rights, or invoking setter functions as the deployer.

-   `function build() public`: Creates the proposal actions and saves them to storage in the proposal contract.

> -   `function simulate() public`: Executes the actions that were previously saved during the `build` step. It's dependent on the successful execution of the `build` function. Without calling `build` first, the `_run` function becomes ineffectual as there would be no predefined actions to execute.

Reviewer 2: ```suggestion
-   `function simulate() public`: Executes the actions that were previously saved during the `build` step. This function's execution is dependent on the successful execution of the `build` function. Without calling `build` first, the `run()` function becomes ineffectual as there would be no predefined actions to execute.
```

> -   `function simulate() public`: Executes the actions that were previously saved during the `build` step. It's dependent on the successful execution of the `build` function. Without calling `build` first, the `_run` function becomes ineffectual as there would be no predefined actions to execute.

Reviewer 2: why do we have both run and simulate? isn't this redundant?

Author: we have both `run` and `simulate`. `_run` has been replaced with `simulate`. I have replaced `_run` with `_simulateActions`

Reviewer 2: ok, so simulate calls all previous functions automatically and then calls run?

Author: `simulte` is called in `run`.

> -   `function validate() public`: Validates the state post-execution. It ensures that the contracts variables and proposal targets are set up correctly.

Reviewer 2: ```suggestion
-   `function validate() public`: Validates system state post-execution. This allows checking that contract's variables and newly deployed contracts are set up correctly.
```

-   `function print() public`: Print proposal description, actions and calldata

The actions in FPS are designed to be loosely coupled for flexible
implementation, with the exception of the build and run functions, which require
sequential execution. This design choice offers developers significant
flexibility and power in tailoring the system to their specific needs. For
example, a developer may choose to only execute the deploy and validate
functions, bypassing the others. This could be suitable in situations where only
initial deployment and final validation are necessary, without the need for
intermediate steps. Alternatively, a developer might opt to simulate a proposal
by executing only the build and run functions, omitting the deploy step if there
is no need to deploy new contracts. FPS empowers developers
with the ability to selectively execute functions based on their unique

> requirements.

Reviewer 2: This ability to pick and choose speeds integration tests, deployment scripts, and governance proposal creation as it becomes easy to access whichever part of a governance proposal that is needed

Author: done

        uint256 balance = Token(token).balanceOf(
            addresses.getAddress("PROTOCOL_TIMELOCK_BRAVO")
        );

        /// CALLS -- mutative and recorded

        // Whitelists the deployed token on the deployed vault.
        Vault(bravoVault).whitelistToken(token, true);

        // Approve the token for the vault.
        Token(token).approve(bravoVault, balance);

        // Deposit all tokens into the vault.
        Vault(bravoVault).deposit(token, balance);
    }
    ```

-   `_validateAction() internal virtual`: It ensures there are no duplicate actions. This method can be further override to include custom checks for a action.

    ```solidity
    function _validateAction(
        address target,
        uint256 value,
        bytes memory data
    ) internal virtual {
        // uses transition storage to check for duplicate actions
        bytes32 actionHash = keccak256(abi.encodePacked(target, value, data));

        uint256 found;

        assembly {
            found := tload(actionHash)
        }

        require(found == 0, "Duplicated action found");

        assembly {
            tstore(actionHash, 1)
        }
    }
    ```

> -   `_validateActions() internal virtual`: This method can be overridden to add custom checks on all actions of proposals. Eg. Number of actions should be 1, First action should be approval etc.

Reviewer: ```suggestion
-   `_validateActions() internal virtual`: This method can be overridden to add custom checks on all actions of proposals. For example, there must be a single action, the first action should be an approval call, etc.
```

    }
    ```

>     Deployments are done by `DEPLOYER_EOA` when running through a proposal `forge script,` and therefore, an address with this exact name must exist in the Addresses.json file. Foundry can be leveraged to actually broadcast the deployments when using the `broadcast` flag combined with `--account` flag. Please refer to the [foundry docs[(https://book.getfoundry.sh/tutorials/best-practices?highlight=broadcast#scripts) for further assistance. On the other hand, when proposals are executed through a `forge script,` the deployer is the proposal contract itself.

Reviewer 2: ```suggestion
    Deployments are done by `DEPLOYER_EOA` when running through a proposal `forge script,` and therefore, an address with this exact name must exist in the Addresses.json file. Foundry can be leveraged to actually broadcast the deployments when using the `broadcast` flag combined with `--account` flag. Please refer to the [foundry docs](https://book.getfoundry.sh/tutorials/best-practices?highlight=broadcast#scripts) for further assistance. Alternatively, when proposals are executed through `forge script,` the deployer address is the proposal contract itself.
```

-   `function afterDeployMock() public`: Post-deployment mock actions. Such actions can include pranking, etching, etc.

<a id="#build-function"></a>

> -   `function build() public`: this function is where most of the FPS magic happens. It utilizes foundry cheat codes to automatically transform plain solidity code into calldata. This calldata can then be proposed on the Governance, signed by the multisig signers, or scheduled on the Timelock. For instance, an action might involve pointing a proxy to a new implementation address after deploying the implementation in the 'deploy' function.

Reviewer 2: ```suggestion
-   `function build() public`: this function is where most of the FPS magic happens. It utilizes foundry cheat codes to automatically transform plain solidity code into calldata encoded for the user's governance model. This calldata can then be proposed on the Governance contract, signed by the multisig signers, or scheduled on the Timelock. For instance, an action might involve pointing a proxy to a new implementation address after deploying the implementation in the 'deploy' function as a privileged admin role in the system.
```

    }
    ```

>     This function is overridden at the proposal-specific contract. Looking at the `run()` function in governor bravo explains how this function works. Here, this function sets the environment for proposal execution and then finally simulates the proposal by calling the `run()` of the parent `Proposal` contract using `super.run()`. In this example, first `primaryForkId` is set to `sepolia`. Next, the `addresses` object is set by reading the `Addresses.json` file. Finally `governor` is set using `setGovernor()` method which is required in `simulate()` and `checkOnChainCalldata()` functions.

Reviewer 2: ```suggestion
    This function is overridden at the proposal-specific contract. The `run()` function in governor bravo explains how this function works. Here, this function sets the environment for proposal execution and then finally simulates the proposal by calling the `run()` of the parent `Proposal` contract using `super.run()`. In this example, first `primaryForkId` is set to `sepolia`. Next, the `addresses` object is set by reading the `Addresses.json` file. Finally `governor` is set using `setGovernor()` method which is required in `simulate()` and `checkOnChainCalldata()` functions.
```
